### DISCUSSION TITLE
help need in righting rule

### DISCUSSION THREAD

**Discussion by robodevchampion** [lf-edge/ekuiper#3134] [Q&A] [answered]
Hi, 

I want to write an rule on top of below message (which is an array of data)

`{
            "sensor1": [
                {
                    "ts": 1724231947811,
                    "values": {
                        "temperature": "200",
                        "humidity": 40.0,
                        "combine": "40:107"
                    }
                },
                {
                    "ts": 1724231947811,
                    "values": {
                        "temperature": "10",
                        "humidity": 41.0,
                        "combine": "41:107"
                    }
                }
            ]
        }`

Now lets say I want to apply a rule where message with temperature < 150 should be filter out, can you please help me in writing SQL for this.

so result would be 
`{"ts":1724231947811,"values":{"combine":"41:107","humidity":41,"temperature":"10"}}`

Thanks

**Comment by ngjaying** [comment]
I think you can use [unnest](https://ekuiper.org/docs/en/latest/sqls/functions/multi_row_functions.html#unnest) to unnest the array row into multiple rows to memory sink and then use [rule pipeline](https://ekuiper.org/docs/en/latest/guide/rules/rule_pipeline.html) to filter each row.

**Reply by robodevchampion** [reply]
ok, one more query to unnest it query would be `SELECT UNNEST(sensor1) FROM tb_test_stream ` 

In my message sensor1 key is basically device_name, so would it be possible to make it dynamic, like always the first/only key , so that it is not required to create different rules for different device, only a single rule would work, so rule would looks like 

 `SELECT UNNEST(*/#some things that extract first key) FROM tb_test_stream ` 
 
 Thanks

**Reply by ngjaying** [reply]
I am afraid there is no dynamic way to unnest now. How many sensor# do you have?

**Reply by robodevchampion** [reply]
well thats ok. we can manage that on gateway level.

I tried to create a rule pipeline 1 rule for unnest and 2nd rule for checking out temperature.

Data In will be` {"device_name":"sensor1","ts":"1724231947811","values":{"combine":"41:107","humidity":41,"temperature":"3"}} `

Need rule for filtering out temperature>30, tried below


```
SELECT * FROM memory_stream 
where json_path_exists(values, "$.temperature > 30")
```

```
SELECT * FROM memory_stream 
where json_path_exists(values, "? @.temperature > 30")
```
Both gives all messages, please check once

Thanks

**Reply by ngjaying** [reply]
how about `where values.temperature > 30`

**Reply by robodevchampion** [reply]
well thats the first try I did, 
```
SELECT * FROM memory_stream 
where values.temperature > 30
```
gives me below error
`'{"error":"run Where error: invalid operation string(200) \\u003e int64(30)"}'`

**Reply by ngjaying** [reply, marked as answer]
`where cast(values.temperature, "bigint") > 30`
